cmd/migrateblobs/main.go

```diff
@@ -0,0 +1,151 @@
+		os.Exit(2)
+	}
+	flag.Parse()
+	if flag.NArg() != 2 {
+		flag.Usage()
+	}
+	if *loggingConfig != "" {
+		if err := loggo.ConfigureLoggers(*loggingConfig); err != nil {
+			fmt.Fprintf(os.Stderr, "cannot configure loggers: %v", err)
+			os.Exit(1)
+		}
+	}
+	if err := run(flag.Arg(0), flag.Arg(1)); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func run(srcConfPath, dstConfPath string) error {
+	logger.Infof("reading configuration")
+	srcConf, err := config.Read(srcConfPath)
+	if err != nil {
+		return errgo.Notef(err, "cannot read config file %q", srcConfPath)
+	}
+	dstConf, err := config.Read(dstConfPath)
+	if err != nil {
+		return errgo.Notef(err, "cannot read config file %q", dstConfPath)
+	}
+
+	logger.Infof("connecting to mongo")
+	session, err := mgo.Dial(srcConf.MongoURL)
+	if err != nil {
+		return errgo.Notef(err, "cannot dial mongo at %q", srcConf.MongoURL)
+	}
+	defer session.Close()
+	db := session.DB("juju")
+
+	cred := &identity.Credentials{
+		URL:        dstConf.SwiftAuthURL,
+		User:       dstConf.SwiftUsername,
+		Secrets:    dstConf.SwiftSecret,
+		Region:     dstConf.SwiftRegion,
+		TenantName: dstConf.SwiftTenant,
+	}
+	client := client.NewClient(cred, dstConf.SwiftAuthMode.Mode, nil)
+	client.SetRequiredServiceTypes([]string{"object-store"})
+	dst := swift.New(client)
+
+	logger.Infof("migrating entity blobs")
+	if err := migrate(db.GridFS("entitystore"), dst, dstConf.SwiftBucket); err != nil {
+		return errgo.Notef(err, "cannot migrate entity blobs")
+	}
+
+	logger.Infof("done")
+	return nil
+}
+
+func migrate(gridfs *mgo.GridFS, sc *swift.Client, dstContainerName string) error {
+	iter := gridfs.Find(nil).Sort("-uploadDate").Iter()
+	defer iter.Close()
+	run := parallel.NewRun(10)
+	var file *mgo.GridFile
+	var counter int32
+	var alreadyExistsCounter int32
+	for gridfs.OpenNext(iter, &file) {
+		// OpenNext close the file before going to the next one
+		file1 := file
+		file = nil
+		run.Do(func() error {
+			defer file1.Close()
+			targetName := strings.TrimPrefix(file1.Name(), "global/")
+			_, err := sc.HeadObject(dstContainerName, targetName)
+			if err == nil {
+				atomic.AddInt32(&alreadyExistsCounter, 1)
+				return nil
+			}
+			if err != nil && !errors.IsNotFound(err) {
+				return errgo.Mask(err)
+			}
+			err = copyObject(file1, sc, dstContainerName)
+			if err != nil {
+				logger.Errorf("cannot copy: %s", err)
+				return errgo.Mask(err)
+			}
+			counter1 := atomic.AddInt32(&counter, 1)
+
+			logger.Infof("%d Migrated %s [%d] %v", counter1, targetName, file1.Size(),
+				file1.UploadDate().Format("2006-01-02 15:04:05"))
+			return nil
+		})
+	}
+	if err := run.Wait(); err != nil {
+		for _, err1 := range err.(parallel.Errors) {
+			logger.Infof("error when migrating", err1)
+		}
+	}
+	if err := iter.Close(); err != nil {
+		return errgo.Notef(err, "cannot iterate entities")
+	}
+
+	logger.Infof("Total entities migrated %d, already existing %d", counter, alreadyExistsCounter)
```

Comment:
We probably want to log this even if we had an error migrating some of the entities.
How about returning counter and alreadyExistsCounter from migrate
and doing this log in the caller?